patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1070 @@
++            case INVULNERABILITY:
++                bypassInvul = true;
++            case ENCHANTMENT:
++                bypassEnchant = true;
++        }
++        return (T) this;
++    }
++
++    @NotNull
++    public T damageHelmetArmor(boolean flag) {
++        damageHelmet = flag;
++        return (T) this;
++    }
++
++    @NotNull
++    public T scalesWithDifficulty(boolean scale) {
++        scalesWithDifficulty = scale;
++        return (T) this;
++    }
++
++    @NotNull
++    public T setFoodExhaustion(float exhaustion) {
++        this.exhaustion = exhaustion;
++        return (T) this;
++    }
++
++    @NotNull
++    public T setDeathMessageCallback(@NotNull Function<LivingEntity, Component> fn) {
++        deathMessageFn = fn;
++        return (T) this;
++    }
++
++    @NotNull
++    public abstract B build();
++}
+diff --git a/src/main/java/io/papermc/paper/entity/damage/builder/AbstractEntityDamageOriginBuilder.java b/src/main/java/io/papermc/paper/entity/damage/builder/AbstractEntityDamageOriginBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..07fd94f5021c246795574ce4604e761cb8147b1d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/damage/builder/AbstractEntityDamageOriginBuilder.java
+@@ -0,0 +1,21 @@
++package io.papermc.paper.entity.damage.builder;
++
++import org.bukkit.entity.Entity;
++import org.jetbrains.annotations.NotNull;
++
++public abstract class AbstractEntityDamageOriginBuilder<T, B> extends AbstractDamageOriginBuilder<T, B> {
++
++    protected final Entity source;
++    protected boolean thorns;
++
++    public AbstractEntityDamageOriginBuilder(@NotNull String name, @NotNull Entity source) {
++        super(name);
++        this.source = source;
++    }
++
++    @NotNull
++    public T setThorns(boolean isThorns) {
++        this.thorns = isThorns;
++        return (T) this;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/entity/damage/builder/DamageOriginBuilder.java b/src/main/java/io/papermc/paper/entity/damage/builder/DamageOriginBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..44541fb0c3b80cde50d3da68ba0b8d5252392917
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/damage/builder/DamageOriginBuilder.java
+@@ -0,0 +1,118 @@
++package io.papermc.paper.entity.damage.builder;
++
++import io.papermc.paper.entity.damage.DamageBypass;
++import io.papermc.paper.entity.damage.DamageOrigin;
++import net.kyori.adventure.text.Component;
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.LivingEntity;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public class DamageOriginBuilder extends AbstractDamageOriginBuilder<DamageOriginBuilder, DamageOrigin> {
++
++    public DamageOriginBuilder(@NotNull String name) {
++        super(name);
++    }
++
++    @NotNull
++    public DamageOrigin build() {
++        return new DamageOrigin() {
```

Comment:
Yeah so basically you want to make it so you can't make custom damage origins by implementing this. All damage origins must be backed by an NMS object. (There is relevant conversation for this)